packages/core/versioning/core-version-http-server/src/example.ts

```diff
@@ -0,0 +1,46 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import { schema } from '@kbn/config-schema';
+import type { IRouter, RequestHandlerContextBase } from '@kbn/core-http-server';
+import type { VersionHTTPToolkit } from './version_http_toolkit';
+
+interface MyCustomContext extends RequestHandlerContextBase {
+  fooService: { create: (value: string) => Promise<void> };
+}
+const vtk = {} as unknown as VersionHTTPToolkit;
+const router = {} as unknown as IRouter<MyCustomContext>;
+
+const versionedRouter = vtk.createVersionedRouter({ router });
+
+// @ts-ignore unused variable
+const versionedRoute = versionedRouter
+  .post({
+    path: '/api/my-app/foo/{name?}',
+    options: { timeout: { payload: 60000 } },
+  })
+  // First version of the API, accepts { foo: string } in the body
+  .addVersion(
+    { version: '1', validate: { body: schema.object({ foo: schema.string() }) } },
+    async (ctx, req, res) => {
+      await ctx.fooService.create(req.body.foo);
+      return res.ok({ body: { foo: req.body.foo } });
+    }
+  )
+  // Second version of the API, accepts { fooName: string } in the body
+  .addVersion(
+    {
+      version: '2',
+      path: '/api/my-app/foo/{id?}', // Update the path to something new
```

Comment:
> If someone still uses the "old" path with a name as the query param, and the route has changed only to accept an optional id, would we not consider that a breaking change and, therefore, not permitted?

Yes, that is a good point. I've updated the example to reflect this as a breaking change.

> My understanding is that we want public HTTP API's to be as stable as possible and any iterations on the route declaration should still work with older versions. Maybe I'm missing some subtlety here.

Also a good point :). My reasoning was that we don't want breaking changes such that a later version removes an earlier version. But I think there are nuanced cases of breaking changes that we should consider allowing if they bring code/functionality to a place that is more maintainable.

CC @rudolf

Reply:
Maybe the example I shared in https://github.com/elastic/kibana/pull/151596#discussion_r1113280567 is helpful.

We want any given version of the API to be stable, but we don't want to hold back teams. If they deem it necessary / beneficial to make a breaking change they should create a new version for that so that they're unblocked.

Reply:
> make a breaking change they should create a new version

Creating a new version for a new path makes me far more comfortable!